### ISSUE TITLE
bpe_span option 

### ISSUE THREAD

**Issue by jiminsun** [octanove/shiba#3]
Hi, 
Thanks for developing and sharing the pre-training code for CANINE!

I'm currently working on pre-training the CANINE model using a Swahili corpus.
First of all, I fixed the `preprocess.py` to modify the Japanese-specific preprocessing methods to Swahili.
As the `bpe_span` option seemed to correspond to the CANINE-S (subword masking) in the original paper, I ran the pre-training with the following flags:

```
python training/train.py \
--data "all_examples.jsonl" \
--logging_steps 50 \
--max_steps 100000 \
--evaluation_strategy steps \
--save_strategy steps \
--eval_steps 1000 \
--save_steps 1000 \
--learning_rate 0.0004 \
--adam_beta2 0.98 \
--adam_epsilon 1e-06 \
--dropout 0.1 \
--weight_decay 0.01  \
--output_dir ./runs/pretrained_shiba \
--masking_type bpe_span \
--gradient_accumulation_steps 8 --per_device_eval_batch_size 4 --per_device_train_batch_size 4
```

However, this resulted in an Exception error:
```
  File "/home/ubuntu/anaconda3/envs/bert/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 47, in fetch
    return self.collate_fn(data)
  File "/home/ubuntu/jimin/shiba/training/masking.py", line 258, in __call__
    input_ids, labels, masked_indices = bpe_span_mask(padded_batch['input_ids'],
  File "/home/ubuntu/jimin/shiba/training/masking.py", line 90, in bpe_span_mask
    bpe_split_string = bpe_tokenizer.encode(original_string)
Exception: Unk token `` not found in the vocabulary
```
Did you face any problems when you ran the model with bpe_span, or would you have any tips on what might have gone wrong / where I should look into to fix the problem?
I did not change the `bpe_vocab.json` file as well. Would this be necessary?

Also, when I applied the `to_examples.py` file my preprocessed file,
the dataset size decreases from around 676993 to 26201, which seems too drastic even considering the length filters.
I wonder if this is a usual outcome that shiba had with the original Japanese dataset.

Thanks in advance!

**Comment by Mindful** [comment]
Hi,
Thanks for using SHIBA!

In regards to your question, BPE span masking does indeed correspond roughly to CANINE's subword masking. 
The issue here is that BPE span masking/subword masking is done using a pretrained tokenizer - the one that ships with SHIBA was pretrained on Japanese Wikipedia, so any characters that don't show up there are going to be out of vocabulary for the tokenizer and consequently lead to errors in subword masking.

The relevant code is [here](https://github.com/octanove/shiba/blob/54ddbf872286660b84f3434d003b34e662b7776d/training/masking.py#L236), but basically if you want to use BPE masking on a Swahili corpus, you need to train a BPE tokenizer on Swahili. Doing something like manually expanding the vocabulary by adding to the `bpe_vocab.json` file is probably _possible_, but unless you actually train a tokenizer it won't learn any sensible subwords in your target language (Swahili), so there wouldn't be much point to using BPE masking.

Fortunately, the tokenizers library from HuggingFace makes it pretty straightforward to train tokenizers: the [quicktour](https://huggingface.co/docs/tokenizers/python/latest/quicktour.html) actually includes a little demo for training a BPE tokenizer. You will also need to tweak the code in RandomSpanMaskingDataCollator to use your newly trained tokenizer, but that should be pretty straightfoward.

Hope this helps, and let me know if you have any other questions.

**Comment by jiminsun** [comment]
Hi, 

Thank you so much for the detailed and quick response, it was super helpful!
I have built a new bpe vocabulary based on my corpus and it works well now.

I just have one question regarding the different options of masking.
In the original CANINE paper, there were **CANINE-S (Subword loss)** and **CANINE-C (Autoregressive character loss)**.
From what I've understood, `bpe_span` is most similar to CANINE-S, and `rand_span` is masking a random span but *not relying on any pre-trained bpe's*. 
In case of `rand_span`, is the masked span predicted auto-regressively, as the method of CANINE-C?

**Comment by Mindful** [comment]
> I have built a new bpe vocabulary based on my corpus and it works well now.

Glad to hear it!

> In case of rand_span, is the masked span predicted auto-regressively, as the method of CANINE-C?

So I just realized that this is a little confusing, and I should probably clarify. SHIBA doesn't actually include a training implementation that exactly matches CANINE-S; what I mean by this is that `bpe_span` masks spans of characters in the same way that CANINE-S masked subwords, but no matter what masking you use SHIBA is always predicting characters. Basically, if you use `bpe_span` you are predicting characters in masked sections determined by subwords (in essence the masking is like CANINE-S but the prediction is not). 

That said, because the prediction style is basically always CANINE-C, everything is autoregressive. Training will be autoregressive no matter what masking type you use. That code is [here](https://github.com/octanove/shiba/blob/main/shiba/model.py#L395) for reference.

Finally, as a side note, I believe `bpe_span` was slower than the other two masking types for us. This may not matter at all for you, but if you find training is slow/GPU utilization is bad, keep an eye out for the possibility of dynamic masking creating a bottleneck.

**Comment by jiminsun** [comment]
Got it, thank you very much for the clarification!
